jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java

```diff
@@ -520,18 +533,85 @@ public String toString()
                     inUse++;
             }
 
-            return String.format("%s{capacity=%d,inuse=%d(%d%%)}",
+            return String.format("%s{capacity=%d,in-use=%d/%d,pooled/acquires=%d/%d,non-pooled/evicts/removes/releases=%d/%d/%d/%d}",
                 super.toString(),
-                _capacity,
+                getCapacity(),
                 inUse,
-                entries > 0 ? (inUse * 100) / entries : 0);
+                entries,
+                _pooled.get(),
+                _acquires.get(),
+                _nonPooled.get(),
+                _evicts.get(),
+                _removes.get(),
+                _releases.get()
+            );
+        }
+
+        private class CompoundBucket extends CompoundPool<RetainableByteBuffer>
+        {
+            private CompoundBucket(ConcurrentBucket concurrentBucket, QueuedBucket queuedBucket)
+            {
+                super(concurrentBucket, queuedBucket);
+            }
+
+            private Pool.Entry<RetainableByteBuffer> evict()
+            {
+                Entry<RetainableByteBuffer> entry = getSecondaryPool().acquire();
+                if (entry == null)
+                    entry = getPrimaryPool().acquire();
+                return entry;
+            }
+        }
+
+        private class ConcurrentBucket extends ConcurrentPool<RetainableByteBuffer>
+        {
+            private ConcurrentBucket(StrategyType strategyType, int maxSize, ToIntFunction<RetainableByteBuffer> maxMultiplex)
+            {
+                super(strategyType, maxSize, maxMultiplex);
+            }
+
+            @Override
+            protected void onAcquired(Entry<RetainableByteBuffer> entry)
+            {
+                super.onAcquired(entry);
+                updateMemory(-_capacity, entry.getPooled().isDirect());
+            }
+
+            @Override
+            protected void onReleased(Entry<RetainableByteBuffer> entry)
+            {
+                super.onReleased(entry);
+                updateMemory(_capacity, entry.getPooled().isDirect());
+            }
```

Comment:
I'd prefer the memory accounting to be done directly in the ABBP methods `acquire`, `reserve` and `release` rather than as side effects of the Pool acquire/release.

Reply:
+1 identical behavior could be achieved by testing the return value of acquire and release, there is no need for any new API.

Reply:
Done.